### ISSUE TITLE
Liftoff

### ISSUE THREAD

**Issue by camisowers** [angelolab/SpaceCat#1]
*This is for internal use only; if you'd like to open an issue or request a new feature, please open a bug or enhancement issue*

### Instructions
This document should be filled out prior to embarking on any project that will take more than a couple hours to complete. The goal is to make sure that everyone is on the same page for the functionality and requirements of new features. Therefore, it's important that this is detailed enough to catch any misunderstandings beforehand. For larger projects, it can be useful to first give a high-level sketch, and then go back and fill in the details. For smaller ones, filling the entire thing out at once can be sufficient.  

**Relevant background**

We would like to generalize the feature analysis scripts in the TNBC_python_scripts [repo](https://github.com/angelolab/TNBC_python_scripts/tree/main/python_files).

Requiring the input cell table in anndata format will speed up the large file loading and allow us to store the output data frames more efficiently.

**Design overview**

The first step would be creating functions to generate features based on simply the cell table, format them appropriately, and create a final feature data frame.

The next step would be allowing for incorporation of other statistics calculated outside of the pipeline (ark spatial stats, fiber stats, etc.). 

**Code mockup**

```
# helper functions for cell table
def create_single_positive_table(adata_cell_table, threshold_list):
   # create binary functional marker table, append to anndata table

def create_double_positive_table(adata_cell_table):
   # pairwise marker thresholding to determine cells positive for combinations of any two markers

def create_functional_tables(adata_cell_table, threshold_list):
   sp_table = create_single_positive_table(adata_cell_table, threshold_list)
   functional_table = create_double_positive_table(sp_table)
   return functional_table
```

```
# main feature generation class
class FeatureSpace:
   feature_data_list = []
   combined_feature_data = None
   combined_feature_data_filtered = None

    
   # HELPER FUNCTIONS
   def __init__(self, adata_table, cluster_columns, compartment_column):
      self.adata_table = adata_table
      self.cluster_columns = cluster_columns
      self.compartment_column = compartment_column
      self.compartment_list = np.unique(adata_table.obs[compartment_column])

   def filter_stats(min_cell_count=5):
      # remove features for cell populations less than 5 cells

   def deduplicate_stats():
      # remove any highly related features

   def remove_correlated_features():
      # filter out features that are highly correlated in compartments

   def get_params_list(freq=True, count=True):
      parms = []
      for cluster_col in self.cluster_columns:
         if freq:
            parms.append([f'{cluster_col}_freq', cluster_col, True])
         if count:
            parms.append([f'{cluster_col}_count', cluster_col, False])

   def cluster_df_helper(cell_table, cluster_col_name, result_name, normalize=False):

   def create_long_df(table, cluster_col_name, result_name, var_name, subset_col=None, normalize=False, drop_cols=None):
      if filter:
         filter_stats()
      if deduplicate:
         deduplicate_stats()
     cluster_df_helper()
   
   def format_helper(compartment_df, col_name, feature_type=None, feature_details=None, cell_types=None):
      if compartment == 'all':
         compartment_df['feature_name_unique'] =  compartment_df['feature_name']
      else:
         compartment_df['feature_name_unique'] = compartment_df['feature_name'] + '__' + compartment

       compartment_df['compartment'] = compartment
       compartment_df['cell_pop_level'] = cell_pop_level
       if cell_types:
          compartment_df['feature_type'] = 'density_ratio'
          compartment_df['feature_type_detail'] = cell_types[0]
          compartment_df['feature_type_detail_2'] = cell_types[1]
       else:
          compartment_df['feature_type'] = feature_type if feature_type != None else col_name
          compartment_df['feature_type_detail'] = compartment_df[col_name] if feature_details == None else feature_details[0]
          compartment_df['feature_type_detail_2'] = '' if feature_details == None else feature_details[1]
          compartment_df = compartment_df[['fov', 'value', 'feature_name','feature_name_unique','compartment', 'cell_pop',
                                         'cell_pop_level', 'feature_type', 'feature_type_detail', 'feature_type_detail_2']]
       return compartment_df
 
   # MAIN FUNCTIONS
   def generate_cluster_stats(cell_table, cluster_df_params):
      cluster_dfs = []
      for result_name, cluster_col_name, normalize in cluster_df_params:
         cluster_dfs.append(create_long_df(cell_table, cluster_col_name, result_name, var_name='cell_type', subset_col='compartment', normalize=normalize))

      # calculate total number of cells per image
      ...
      # calculate total number of cells per region per image
      ...
      # calculate proportions of cells per region per image
      ...
      # create single df with appropriate metadata
      total_df_clusters = pd.concat(cluster_dfs, axis=0)
      adata_table. uns['cluster_stats'] = total_df

   def generate_stats(table, params, drop_cols, df_name, var_name, filter=True, deduplicate=True):
      stats_dfs = []
      for result_name, cluster_col_name, normalize in params:
         stats_dfs.append(create_long_df(table=table,
                                                 result_name=result_name,
                                                 var_name='functional_marker'
                                                 cluster_col_name=cluster_col_name,
                                                 var_name=var_name,
                                                 drop_cols=drop_cols,
                                                 normalize=True,
                                                 subset_col='tumor_region'))
      total_df = pd.concat(stats_dfs, axis=0)
      adata_table. uns[df_name] = total_df

     # save filtered version as well
     if filter:
        total_df_filtered = filter_stats(total_df)
     if deduplicate:
        total_df_filtered = deduplicate_stats(total_df_filtered)
     adata_table. uns[df_name+'_filtered'] = total_df_filtered

   def generate_high_level_stats(stats_df, broad_cluster_col, minimum_density = 0.0005):
      # compute abundance of cell types
      abundance_features = [[f'{broad_cluster_col}_density', 'cluster_broad_density', 'broad'],
                                               ['total_cell_density', 'total_density', 'broad']]

      for cluster_name, feature_name, cell_pop_level in abundance_features:
         input_df = cluster_df_core[cluster_df_core['metric'].isin([cluster_name])]
         for compartment in compartment_list:
            compartment_df = input_df[input_df.subset == compartment].copy()
            compartment_df['feature_name'] = compartment_df.cell_type + '__' + feature_name
          
            compartment_df_formatted = format_helper(compartment_df, col_name='cell_type', feature_type='density')

       # compute ratio of broad cell type abundances
       input_df = cluster_df_core[cluster_df_core['metric'].isin([f'{broad_cluster_col}_density'])]
       for compartment in compartment_list:
          compartment_df = input_df[input_df.subset == compartment].copy()
          cell_types = compartment_df.cell_type.unique()

          for cell_type1, cell_type2 in itertools.combinations(cell_types, 2):
             cell_type1_df = compartment_df[compartment_df.cell_type == cell_type1].copy()
             cell_type2_df = compartment_df[compartment_df.cell_type == cell_type2].copy()

             # only keep FOVS with at least one cell type over the minimum density
             cell_type1_mask = cell_type1_df.value > minimum_density
             cell_type2_mask = cell_type2_df.value > minimum_density
             cell_mask = cell_type1_mask.values | cell_type2_mask.values
             cell_type1_df = cell_type1_df[cell_mask]
             cell_type2_df = cell_type2_df[cell_mask]

             # add minimum density to avoid log2(0)
             cell_type1_df['ratio'] = np.log2((cell_type1_df.value.values + minimum_density) /
                                         (cell_type2_df.value.values + minimum_density))
             cell_type1_df['value'] = cell_type1_df.ratio.values
             cell_type1_df['feature_name'] = cell_type1 + '__' + cell_type2 + '__ratio'

             cell_type1_df_formatted = format_helper(compartment_df, col_name=None, cell_types=(cell_type1, cell_type2))
             self.feature_data_list.append(cell_type1_df_formatted)
         
      ## on hold for now: compute ratio of specific cell type abundances
      ## on hold for now: compute proportion of cells in a given cell type
      ## on hold for now: compute diversity of different levels of granularity

   def format_features(stats_df, col_name, metrics):
      # takes in an input_df, and for each compartment, adds necessary feature information 
      for  metric_name, cell_pop_level in metrics:
         input_df = stats_df[stats_df['metric'].isin([metric_name])]
         for compartment in compartment_list:
            compartment_df = input_df[input_df.subset == compartment].copy()
            compartment_df['feature_name'] = compartment_df[col_name]+ '+__' + compartment_df.cell_type

            compartment_df_formatted = format_helper(compartment_df, col_name)
            self.feature_data_list.append(compartment_df_formatted)


   def combine_features(correlation_filtering=0.7):
      feature_df = pd.concat(self.feature_data_list)
      feature_df = feature_df.rename(columns={'value': 'raw_value'})
      ...
      self.combined_feature_data = feature_df
      adata_table. uns['combined_feature_data'] = feature_df
      feature_df.to_csv()

      if correlation_filtering:
         feature_df_filtered = remove_correlated_features(feature_df)
         self.combined_feature_data_filtered = feature_df_filtered
         adata_table. uns['combined_feature_data_filtered'] = feature_df
         feature_df_filtered.to_csv()

      # save feature metadata
      adata_table. uns['feature_metadata'] = feature_metadata_df
      feature_metadata_df.to_csv()

```

```
# notebook script

# preprocessing 
adata_table = anndata.read_h5ad(os.path.join(base_directory, 'adata_table.h5ad'))
threshold_list = ['Ki67', 0.002], ['CD38', 0.004], ...]
adata_table.uns['mark_positivity'] = create_functional_tables(adata_table, threshold_list) # can probably find a better place than uns for this

cluster_columns = []
compartment_list = []
features = FeatureSpace(adata_table, cluster_columns, compartment_list)


# Generate counts and proportions of cell clusters per FOV
params = features.get_params_list(freq=True, count=True)
cell_table_clusters = adata_table[[]] # subset for cluster data
features.generate_cluster_stats(cell_table_clusters, params)

# Create summary dataframe with proportions and counts of different functional marker populations
params = features.get_params_list(freq=True, count=True)
cell_table_func = adata_table[[]] # subset for functional data
features.generate_stats(cell_table_func, params=params, df_name='functional_stats', var_name='functional_marker')

# morphology metric summary
params =  get_params_list(freq=True, count=False)
cell_table_morph = adata_table[[]] # subset for morphological data
features.generate_stats(cell_table_morph, params=params, df_name='morphology_stats', var_name='morphology')

# can also be done for diversity & linear distance stats
...

# format previous features
feature_set = [[features.adata_table.uns['functional_stats'], 'functional_marker'], 
                        [features.adata_table.uns['morphology_stats'], 'morphology']]   # can add diversity and  linear distance       
metrics = [['cluster_freq', 'med'], ['total_freq', 'broad']]
for stats_df, col_name in feature_set:
   features.format_features(stats_df, col_name, metrics)


# high level cluster stats
stats_df = features.adata_table.uns['cluster_stats']
broad_cluster_col = '' # will generate all pairwise ratios for this clustering level
features.generate_high_level_stats(stats_df, broad_cluster_col)

# combine features
features.combine_features()

```

Aggregate feature **calculation** function:
 - `create_long_df()`: summarize a statistic by cluster, with the option to subset by an additional feature (compartment)
    - based on [create_long_df_by_cluster](https://github.com/angelolab/TNBC_python_scripts/blob/fdd0ad992d1dd098752877ab166e35c63660b05c/python_files/utils.py#L105C74-L105C84) and [create_long_df_by_functional](https://github.com/angelolab/TNBC_python_scripts/blob/fdd0ad992d1dd098752877ab166e35c63660b05c/python_files/utils.py#L181C5-L181C33)
    - this will generate cluster_df_per_core.csv, functional_df_per_core.csv, morph_df_per_core.csv (saved to anndata instead of csv)
    - **missing**: ( diversity_df_per_core.csv, distance_df_per_core.csv, mixing score, kmeans, fiber, ecm stats)

High level feature **calculation** function:
- `generate_high_level_stats()`: summarizes stats using various output files created by `create_long_df()`
   - generate stats for cell type diversity across image, ratio of cell abundances, proportions of intermediate cell types in broad 
   - pulls from [6_create_fov_stats.py](https://github.com/angelolab/TNBC_python_scripts/blob/fdd0ad992d1dd098752877ab166e35c63660b05c/python_files/postprocessing/6_create_fov_stats.py)
 
Feature **formatting** function:
- `format_features()`: reformats dfs from the resulting from `create_long_df()` to have useful metadata

Feature **combining** function: 
- `combine_features()`: combines the store dfs into a single timepoint_features.csv
   - also generates df with correlated features removed
   - [code](https://github.com/angelolab/TNBC_python_scripts/blob/fdd0ad992d1dd098752877ab166e35c63660b05c/python_files/postprocessing/6_create_fov_stats.py#L666)

**Required inputs**

Cell table, annotations_by_mask (assigns cells to compartments in the image),  harmonized metadata.
 - this can all be stored in single anndata table

**Output files**
timepoint_features.csv, timepoint_features_filtered.csv, feature_metadata.csv

**Timeline**
Give a rough estimate for how long you think the project will take. In general, it's better to be too conservative rather than too optimistic. 
- [ ] A couple days
- [ ] A week
- [x] Multiple weeks. For large projects, make sure to agree on a plan that isn't just a single monster PR at the end. 

Estimated date when a fully implemented version will be ready for review:
PR #2 Friday 8/16 (starter class)
PR #3 Friday 8/23 (functional)

Estimated date when the finalized project will be merged in: 
Friday 8/30

**Comment by camisowers** [comment]
@jranek this is a monster class composed of pseudocode and real code lol. Feel free to skim over it but I specifically would appreciate your thoughts on the best format to save the data frames within the anndata object, as well as if the layout of the class is confusing.

**Comment by ngreenwald** [comment]
This looks awesome, can discuss today. A couple points to talk about:
1. How easy is it to convert a .csv into an anndata table, assuming the columns have been named appropriately? I imagine most people will have something that is easy to convert to a .csv file. If we can get them to do the renaming/reformatting so the table is correct, I think it would be nice if we included anndata conversion logic as part of the tool. 
2. What level of control does it make sense to give users over which cluster resolutions and statistics are generated for which features? Currently, this is all basically hard-coded (i.e. densities at intermediate and broad level, functional marker at intermediate, ratios at broad, custom ratios at intermediate, proportions only within lineages and never global)
3. Are the different levels of feature details useful, or should this be simplified 
4. What is the best way to organize this so that users have the option to 1) change nothing, and get default decisions for basically everything and 2) have the ability to customize specific thresholds with some degree of control. Ideally, if they want to change nothing, it could be only a couple lines of code to run the whole thing, which would mean having a top-level function to call most of what's currently in the notebook. The downside is then any params need to be passed through that.

**Comment by jranek** [comment]
Alternative structural ideas:
- Append each fov-level statistic to `adata.uns` prior to melting and combining the dataframes
```python
adata.uns['cell_cluster_abundance'] 
adata.uns['cell_cluster_broad_abundance'] 
...
```
- Create MuData object to house individual anndata objects 
    - anndata object for the cell table (i.e. `adata.X` containing expression, `adata.obs` containing cell-level attributes + morphology + patient level metadata)
    - anndata object for final engineered features

**Comment by camisowers** [comment]
Closed by PRs #2, #3, and #4. Morphology features excluded for now.
